ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/KsqlRestConfig.java

```diff
@@ -183,4 +220,170 @@ public KsqlRestConfig(final Map<?, ?> props) {
+      final Function<URL, Integer> portResolver,
+      final Logger logger
+  ) {
+    return getString(INTER_NODE_LISTENER_CONFIG).equals(USE_FIRST_LISTENER)
+        ? getInterNodeListenerFromFirstListener(portResolver, logger)
+        : getInterNodeListenerFromExplicitConfig(logger);
+  }
+
+  private URL getInterNodeListenerFromFirstListener(
+      final Function<URL, Integer> portResolver,
+      final Logger logger
+  ) {
+    final List<String> configValue = getList(LISTENERS_CONFIG);
+
+    final URL firstListener = parseUrl(configValue.get(0), LISTENERS_CONFIG);
+
+    final InetAddress address = parseInetAddress(firstListener.getHost())
+        .orElseThrow(() -> new ConfigException(
+            LISTENERS_CONFIG,
+            configValue,
+            "Could not resolve first host"
+        ));
+
+    final URL listener = sanitizeInterNodeListener(
+        firstListener,
+        portResolver,
+        address.isAnyLocalAddress()
+    );
+
+    logInterNodeListener(
+        logger,
+        listener,
+        Optional.of(address),
+        "first '" + LISTENERS_CONFIG + "'"
+    );
+
+    return listener;
+  }
+
+  private URL getInterNodeListenerFromExplicitConfig(final Logger logger) {
+    final String configValue = getString(INTER_NODE_LISTENER_CONFIG);
+
+    final URL listener = parseUrl(configValue, INTER_NODE_LISTENER_CONFIG);
+
+    if (listener.getPort() <= 0) {
+      throw new ConfigException(INTER_NODE_LISTENER_CONFIG, configValue, "Must have valid port");
+    }
+
+    // Valid for address to not be resolvable, as may be _externally_ resolvable:
+    final Optional<InetAddress> address = parseInetAddress(listener.getHost());
+
+    address.ifPresent(a -> {
+      if (a.isAnyLocalAddress()) {
+        throw new ConfigException(INTER_NODE_LISTENER_CONFIG, configValue, "Can not be wildcard");
+      }
+    });
+
+    logInterNodeListener(
+        logger,
+        listener,
+        address,
+        "'" + INTER_NODE_LISTENER_CONFIG + "'"
+    );
+
+    return listener;
+  }
+
+  private static void logInterNodeListener(
+      final Logger logger,
+      final URL listener,
+      final Optional<InetAddress> address,
+      final String sourceConfigName
+  ) {
+    address.ifPresent(a -> {
+      if (a.isLoopbackAddress()) {
+        logger.warn(
+            "{} config is set to a loopback address: {}. Intra-node communication will only work "
+                + "between nodes running on the same machine.",
+            sourceConfigName, listener
+        );
+      }
+
+      if (a.isAnyLocalAddress()) {
+        logger.warn(
+            "{} config uses wildcard address: {}. Intra-node communication will only work "
+                + "between nodes running on the same machine.",
+            sourceConfigName, listener
+        );
+      }
+    });
+
+    logger.info("Using {} config for intra-node communication: {}", sourceConfigName, listener);
+  }
+
+  private static URL sanitizeInterNodeListener(
```

Comment:
Add some comments indicating that this is only used for the "first listener" which may use the anyLocalAddress symbolic address.